File: http/source/crypto_route_map.c
#include "crypto_route_map.h"

#include <string.h>

http3_crypto_grade_t http3_grade_for_risk(uint8_t measured_risk)
{
    if (measured_risk >= 3U) {
        return HTTP3_CRYPTO_CR4;
    }
    if (measured_risk == 2U) {
        return HTTP3_CRYPTO_CR3;
    }
    if (measured_risk == 1U) {
        return HTTP3_CRYPTO_CR2;
    }
    return HTTP3_CRYPTO_CR1;
}

static http3_crypto_grade_t max_grade(http3_crypto_grade_t a,
                                       http3_crypto_grade_t b)
{
    return (a > b) ? a : b;
}

int http3_route_evaluate(const http3_route_policy_t *policy,
                         const http3_route_node_t *nodes,
                         size_t node_count,
                         http3_route_decision_t *decision,
                         http3_crypto_grade_t *country_capsule_grades,
                         size_t capsule_capacity)
{
    size_t i;
    http3_crypto_grade_t grade;
    size_t capsule_count = 0U;

    if (policy == NULL || nodes == NULL || decision == NULL || node_count == 0U) {
        return -1;
    }

    grade = policy->minimum_grade;
    grade = max_grade(grade, policy->source_grade);
    grade = max_grade(grade, policy->destination_grade);

    for (i = 0U; i < node_count; ++i) {
        http3_crypto_grade_t risk_grade = http3_grade_for_risk(nodes[i].measured_risk);
        grade = max_grade(grade, risk_grade);

        if (policy->allow_jurisdictional_capsules &&
            nodes[i].emergency_capsule_allowed &&
            country_capsule_grades != NULL &&
            capsule_count < capsule_capacity) {
            country_capsule_grades[capsule_count++] = risk_grade;
        }
    }

    if (policy->strengthen_after_threshold &&
        node_count > 0U &&
        (node_count - 1U) > policy->hop_threshold) {
        grade = max_grade(grade, HTTP3_CRYPTO_CR3);
    }

    decision->grade = grade;
    decision->hop_count = node_count - 1U;
    decision->capsule_country_count = capsule_count;
    decision->reason =
        "Protection is monotonic: route conditions may strengthen the selected grade but do not silently downgrade it.";

    return 0;
}

File: http/source/crypto_route_map.py
"""HTTP 3.0 Cryptographic Route Map reference model.

This module models route-aware cryptographic policy without performing packet
interception or exposing sensitive identity data. It is intentionally a policy
layer: QUIC/HTTP/3 packet processing remains the responsibility of the actual
transport implementation.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum
from typing import Iterable, Sequence


class CryptoGrade(IntEnum):
    """Cryptographic protection policy level."""

    CR1 = 1
    CR2 = 2
    CR3 = 3
    CR4 = 4


@dataclass(frozen=True)
class RouteNode:
    """One logical country/network point in a route."""

    node_id: str
    country_code: str
    technology_grade: int = 0
    socialism_standard: str = "S-N/A"
    measured_risk: int = 0
    emergency_capsule_allowed: bool = False

    def __post_init__(self) -> None:
        if not self.node_id:
            raise ValueError("node_id must not be empty")
        if not (0 <= self.technology_grade <= 4):
            raise ValueError("technology_grade must be between 0 and 4")
        if not (0 <= self.measured_risk <= 4):
            raise ValueError("measured_risk must be between 0 and 4")


@dataclass(frozen=True)
class RoutePolicy:
    """Policy inputs used to select cryptographic protection."""

    source_grade: CryptoGrade = CryptoGrade.CR2
    destination_grade: CryptoGrade = CryptoGrade.CR2
    minimum_grade: CryptoGrade = CryptoGrade.CR2
    hop_threshold: int = 4
    strengthen_after_threshold: bool = True
    allow_jurisdictional_capsules: bool = True

    def __post_init__(self) -> None:
        if self.hop_threshold < 0:
            raise ValueError("hop_threshold must not be negative")


@dataclass(frozen=True)
class RouteDecision:
    """Result of evaluating a route."""

    grade: CryptoGrade
    hop_count: int
    countries: tuple[str, ...]
    capsule_countries: tuple[str, ...] = field(default_factory=tuple)
    reason: str = ""


class CryptographicRouteMap:
    """Calculate a monotonic cryptographic policy for a route.

    The policy never uses a country's political classification as a direct
    security score. Country metadata may inform external governance policy,
    while measured network conditions and explicit minimums control protection.
    """

    def __init__(self, policy: RoutePolicy | None = None) -> None:
        self.policy = policy or RoutePolicy()

    @staticmethod
    def _grade_for_risk(risk: int) -> CryptoGrade:
        return CryptoGrade(max(1, min(4, risk + 1)))

    def evaluate(self, nodes: Sequence[RouteNode]) -> RouteDecision:
        if not nodes:
            raise ValueError("route must contain at least one node")

        hop_count = max(0, len(nodes) - 1)
        risk_grade = max(
            (self._grade_for_risk(node.measured_risk) for node in nodes),
            default=CryptoGrade.CR1,
        )
        grade = max(
            self.policy.minimum_grade,
            self.policy.source_grade,
            self.policy.destination_grade,
            risk_grade,
        )

        if self.policy.strengthen_after_threshold and hop_count > self.policy.hop_threshold:
            grade = max(grade, CryptoGrade.CR3)

        capsule_countries: list[str] = []
        if self.policy.allow_jurisdictional_capsules:
            for node in nodes:
                if node.emergency_capsule_allowed and node.country_code not in capsule_countries:
                    capsule_countries.append(node.country_code)

        return RouteDecision(
            grade=CryptoGrade(grade),
            hop_count=hop_count,
            countries=tuple(node.country_code for node in nodes),
            capsule_countries=tuple(capsule_countries),
            reason=(
                "Protection is monotonic: route conditions may strengthen the "
                "selected grade but do not silently downgrade it."
            ),
        )

    def evaluate_country_path(self, countries: Iterable[str]) -> RouteDecision:
        """Convenience evaluation for a path where only country order is known."""
        nodes = [
            RouteNode(node_id=f"country-{index}", country_code=country)
            for index, country in enumerate(countries)
        ]
        return self.evaluate(nodes)
